qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,698 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Class to store and manage the results of a calibration experiments."""
+
+from collections import namedtuple, defaultdict
+from datetime import datetime
+from typing import Any, Dict, Set, Tuple, Union, List, Optional
+import csv
+import dataclasses
+import regex as re
+
+from qiskit.circuit import Gate
+from qiskit import QuantumCircuit
+from qiskit.pulse import (
+    Schedule,
+    DriveChannel,
+    ControlChannel,
+    MeasureChannel,
+    Call,
+    Play,
+    ShiftPhase,
+    SetPhase,
+    ShiftFrequency,
+    SetFrequency,
+)
+from qiskit.pulse.channels import PulseChannel
+from qiskit.pulse.transforms import inline_subroutines
+from qiskit.circuit import Parameter, ParameterExpression
+from qiskit_experiments.calibration.exceptions import CalibrationError
+from qiskit_experiments.calibration.parameter_value import ParameterValue
+
+ParameterKey = namedtuple("ParameterKey", ["schedule", "parameter", "qubits"])
+
+
+class Calibrations:
+    r"""
+    A class to manage schedules with calibrated parameter values.
+    Schedules are stored in a dict and are intended to be fully parameterized,
+    including the index of the channels. The parameter values are stored in a
+    dict where parameters are keys. This class supports:
+    - having different schedules share parameters
+    - allows default schedules for qubits that can be overridden of specific qubits.
+
+    Parametric channel naming convention.
+    Channels must be name according to a predefined pattern so that self can resolve
+    the channels and control channels when assigning values to the parametric channel
+    indices. This pattern is "^ch\d[.\d]*\${0,1}[\d]*$", examples of which include "ch0",
+    "ch1", "ch0.1", "ch0$", "ch2$3", and "ch1.0.3$2". The "." delimiter is used to
+    specify the different qubits when looking for control channels.
+    The optional $ delimiter is used to specify which control channel to use
+    if several control channels work together on the same qubits. For example, if the
+    control channel configuration is {(3,2): [ControlChannel(3), ControlChannel(12)]}
+    then given qubits (2, 3) the name "ch1.0$1" will resolve to ControlChannel(12) while
+    "ch1.0$0" will resolve to ControlChannel(3).
+
+    Parameter naming restriction.
+    Each parameter must have a unique name within each schedule. For example, it is
+    acceptable to have a parameter named 'amp' in the schedule 'xp' and a different
+    parameter instance named 'amp' in the schedule named 'xm'. It is not acceptable
+    to have two parameters named 'amp' in the same schedule.
+    """
+
+    # The channel indices need to be parameterized following this regex.
+    __channel_pattern__ = r"^ch\d[.\d]*\${0,1}[\d]*$"
+
+    def __init__(self, control_config: Dict[Tuple[int, ...], List[ControlChannel]] = None):
```

Comment:
Perhaps we can take target backend name: str for equality check.

Reply:
Perhaps. I'd like to follow this PR up with one where we do e.g. spectroscopy. This will then make it pretty clear how we interact with the backend.